server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -50,187 +51,267 @@
+    if (queue == null) {
+      log.makeAlert("Tier[%s] has no servers! Check your cluster configuration!", tier).emit();
+      return Collections.emptyList();
+    }
 
-      if (serverQueue == null) {
-        log.makeAlert("Tier[%s] has no servers! Check your cluster configuration!", tier).emit();
+    return queue.stream().filter(predicate).collect(Collectors.toList());
+  }
+
+  @Nullable
+  private static ServerHolder assignPrimary(
+      final Object2IntMap<String> targetReplicants,
+      final DruidCoordinatorRuntimeParams params,
+      final DataSegment segment
+  )
+  {
+    ServerHolder topCandidate = null;
+    for (final Object2IntMap.Entry<String> entry : targetReplicants.object2IntEntrySet()) {
+      final int targetReplicantsInTier = entry.getIntValue();
+      if (targetReplicantsInTier <= 0) {
         continue;
       }
 
-      final int maxSegmentsInNodeLoadingQueue = params.getCoordinatorDynamicConfig()
-                                                      .getMaxSegmentsInNodeLoadingQueue();
+      final String tier = entry.getKey();
 
-      Predicate<ServerHolder> serverHolderPredicate;
-      if (maxSegmentsInNodeLoadingQueue > 0) {
-        serverHolderPredicate = s -> (s != null && s.getNumberOfSegmentsInQueue() < maxSegmentsInNodeLoadingQueue);
-      } else {
-        serverHolderPredicate = Objects::nonNull;
+      final List<ServerHolder> holders = getFilteredHolders(
+          tier,
+          params.getDruidCluster(),
+          createLoadQueueSizeLimitingPredicate(params)
+      );
+      if (holders.isEmpty()) {
+        continue;
       }
 
-      final List<ServerHolder> serverHolderList = serverQueue.stream()
-                                                             .filter(serverHolderPredicate)
-                                                             .collect(Collectors.toList());
-
-      final BalancerStrategy strategy = params.getBalancerStrategy();
-      if (availableSegments.contains(segment)) {
-        int assignedCount = assign(
-            params.getReplicationManager(),
-            tier,
-            totalReplicantsInCluster,
-            expectedReplicantsInTier,
-            totalReplicantsInTier,
-            strategy,
-            serverHolderList,
-            segment
+      final ServerHolder candidate = params.getBalancerStrategy().findNewSegmentHomeReplicator(segment, holders);
+      if (candidate == null) {
+        log.warn(
+            "Not enough [%s] servers or node capacity to assign primary segment[%s]! Expected Replicants[%d]",
+            tier, segment.getIdentifier(), targetReplicantsInTier
         );
-        stats.addToTieredStat(ASSIGNED_COUNT, tier, assignedCount);
-        totalReplicantsInCluster += assignedCount;
+      } else if (topCandidate == null || candidate.getServer().getPriority() > topCandidate.getServer().getPriority()) {
+        topCandidate = candidate;
```

Comment:
This implementation will call the balancer strategy a whole lot more than the prior implementation if I'm reading this correctly. How many times is the expensive balancer code run in this PR vs the old implementation?

Reply:
The implementation will make N - 1 more calls where N is the number of tiers. If we have about 2 tiers per rule, this will make one more call.

Reply:
per segment? so double the number of strategy calls?

Reply:
Is there a way to make it smaller?

Reply:
We can cache the result (i.e., memoization) so that there will not be additional calls.

Reply:
different than https://github.com/druid-io/druid/pull/4731 ?

AKA, would https://github.com/druid-io/druid/pull/4731 be a requirement for this to be performant?

Reply:
To get around the performance hit, maybe a config flag enabling "best" search which defaults to false, and the false case returns a holder on the first tier it finds a good result in?

Reply:
No, what I meant the caching here is much more simpler. Basically, in the original implementation, the "strategy" method will be called _R_ times per tier, where _R_ is the number of replicas for that tier.

In the PR, when primary replica needs to be identified, the strategy method will be called _T_ times, where _T_ is the number of tiers, for the primary replica. From the _T_ results, we chose the most prioritized one to be the primary holder and the other _T - 1_ are effectively discarded. What I am saying is that these discarded results could be reused for the assignment of non-primary replicas.

Reply:
@drcrallen I have committed a change (https://github.com/druid-io/druid/pull/4757/commits/748112c9790623d479b53419650821167b9b8c1a) that does what I've mentioned above.